cep-3.md

```diff
@@ -0,0 +1,109 @@
+<table>
+<tr><td> Title </td><td> Using the Mamba solver in Conda </td>
+<tr><td> Status </td><td> Draft </td></tr>
+<tr><td> Author(s) </td><td> Jannis Leidel &lt;jleidel@anaconda.com&gt;</td></tr>
+<tr><td> Created </td><td> Jul 20, 2021</td></tr>
+<tr><td> Updated </td><td> Jul 22, 2021</td></tr>
+<tr><td> Discussion </td><td> https://github.com/conda/ceps/pulls/2 </td></tr>
+<tr><td> Implementation </td><td> NA </td></tr>
+</table>
+
+## Abstract
+
+[Conda]'s current default dependency solver based on [pycosat] has a number
+of downsides that led to community efforts to write a faster and more
+modern solver based on [libsolv]. The result was the basis for the
+alternative Conda implementation called [Mamba].
+
+The intention of this CEP is to propose an experimental feature for
+Conda to add support for the solver used in [Mamba], based on the plugin
+architecture proposed in [CEP 2].
+
+## Motivation
+
+Conda's functionality and usability is predicated on the runtime
+characteristics of its low-level dependency resolver since it's a
+major part of its architecture -- next to the code infrastructure
+for the command line interface, the environment management and
+the handling of repo and channel data.
+
+Speed, accuracy, debuggability and other impacts on the user
+experience are heavily reliant on the underlying functionality of the
+solver and Conda's ability to interpret and communicate it to the user.
+For example long running calls of the Conda install command to resolve
+complex environments and dependency trees directly impacts users'
+real world experience of the package manager.
+
+In addition, there is a wide-ranging proliferation of community efforts
+to improve the usability issues with Conda outside the regular Conda
+code project which makes it harder to retain a consistent technical
+and community roadmap. As such the experimental integration of the Mamba
+solver is to be considered but the first step to affect lasting change
+in the Conda community and recognize the innovative solutions that have
+been worked on by community members.
+
+## Specificaton
+
+Conda contains a solver registry in the `conda.resolve` module that
+must be extended to use the plugin architecture specified in CEP 2.
+
+Conda currently (4.10.x) supports three solvers:
```

Comment:
One important differentiation as far as I understand `mamba` is also that the building of the SAT clauses is done in C++ and not in Python. This also saves a considerable amount of time and is different to the current interface in https://github.com/conda/conda/blob/248741a843e8ce9283fa94e6e4ec9c2fafeb76fd/conda/resolve.py#L44

We should still be able to provide a registry in the `conda.resolve` module, probably just on a different level than the current one.

Reply:
Yeah, this'll be part of the implementation of the CEP, I'll update the draft to include a note about the difference in API levels and a need to refactor and/or integrate appropriately. Thanks for raising this!

Reply:
@xhochy do you have any data on this? The claim(? -  as far as I understand) that the SAT clause definition is somehow a bottleneck seems a bit weird to me :sweat_smile:

Admittedly, I am not super knowledgeable on SAT solvers, but I would not expect this to take for than a couple milliseconds.

Usually, the bottleneck on dependency resolvers that work on remote sources is fetching the metadata necessary for the dependency resolution. So, it usually tends to be better to give more weight to the resolution algorithm instead of this kind of optimizations :smile:

I haven't looked too much in all the possible backends, I am just raising the question.

Reply:
This meta issue has some useful links to understand the bottlenecks: https://github.com/conda/conda/issues/7700